### DISCUSSION TITLE
Charts axis ticks are not completely visible

### DISCUSSION THREAD

**Discussion by imrankhurram-checkit** [primefaces/community#1856] [PrimeFaces] [answered]
Charts axis ticks are not completely visible as something is drawn on the canvas over the ticks. It looks like this is a bug in Primefaces 12 but is there any workaround? Below are samples:
![image](https://github.com/primefaces/community/assets/64078961/783d8e3d-4004-4ddc-9ef7-856bb853c4d4)
Here you can see most of the x-axis ticks are not visible completely

![image (2)](https://github.com/primefaces/community/assets/64078961/1cbd8206-cf8a-415e-815a-ada17df2caff)
Here you can see the second and third x-axis ticks are not visible completely.

**Comment by melloware** [comment]
See this discussion here: https://github.com/orgs/primefaces/discussions/1832

**Reply by imrankhurram-checkit** [reply]
but that discussion shows issue in Primefaces 6.2. I am already new version i.e.12 which should have the fix, but still not working. Can you show me the code that should come in the patch? so I can integrate that please

**Reply by melloware** [reply]
@imrankhurram-checkit the patch doesn't fix it.  It looks like a core bug in jqPlot which has not had any fixes in 6 years.  I would switch Charting libraries off of the jqPlot lib.

**Reply by imrankhurram-checkit** [reply]
charting in my app is having alot of things which can't be switched that easily. Any other workaround to be done over it?

**Reply by melloware** [reply]
you can report a defect at jqPlot: https://github.com/jqPlot/jqPlot

**Reply by imrankhurram-checkit** [reply]
https://github.com/jqPlot/jqPlot/issues/208
No one is addressing that on jqPlot

**Reply by melloware** [reply]
That is all you can do.  But like i said it hasn't had a commit in 6 years so you should think about another charting library.

**Reply by harbulot** [reply]
Despite the low activity in jqPlot, it looks like I had coincidentally reported the same problem (**with a fix**) just a few days before: jqPlot/jqPlot#204

**Reply by melloware** [reply]
Nice @harbulot that is slick!  You guys should be able to patch your legacy PF JARs with these changes or MonkeyPatch it if its possible.

**Reply by imrankhurram-checkit** [reply]
@harbulot @melloware can you please tell me how to use this patch in my Primefaces? I am unable to find Jqplot libraries there. I am using Primefaces 12

**Reply by melloware** [reply]
@imrankhurram-checkit the jqPlot is located in the PF 12 JAR in `\META-INF\resources\primefaces\charts`

**Reply by imrankhurram-checkit** [reply]
@melloware thanks for the response. At this path, I am seeing three files only
- jqplot.txt
- charts.js
- charts.css
Whereas the file having the fix is jqplot.core.js as follows:
https://github.com/jqPlot/jqPlot/pull/205/files#diff-3c963554f92b4e9e4ac57ebf21bf69f070e18b4c2d451c05677309fe5c2e74bbR285

How to use this file at the above link so that it can replace/overwrite things

**Reply by melloware** [reply]
its in chart.js the chart.js is combined jqplot.js and the widget.js

**Reply by imrankhurram-checkit** [reply]
@melloware then how to use this jqplot.core.js ? I cannot find those lines as well in chart.js which they have provided in jqplot.core.js as a fix.

**Reply by melloware** [reply]
Maybe this will help: https://github.com/primefaces/primefaces/tree/12.0.0/primefaces/src/main/resources/META-INF/resources/primefaces/charts  that is PF 12.0.0 before its compressed into 1 file charts.js

**Reply by harbulot** [reply]
@melloware @imrankhurram-checkit Sorry, I don't use PrimeFaces myself, I only followed this discussion after seeing jqPlot/jqPlot#208.

That said, it seems the main line of interest is [Line 299 in `resources/primefaces/charts/0-jqPlot.js`](https://github.com/primefaces/primefaces/blob/12.0.0/primefaces/src/main/resources/META-INF/resources/primefaces/charts/0-jqPlot.js#L299).

### Option 1

As the most basic fix, you could change line 299 like this:

```diff
-             var cctx = canvas.getContext('2d');
+             var cctx = canvas.getContext('2d', { willReadFrequently: true });
```

This is fixed and not configurable.

### Option 2

You could do something like the patch I've suggested in jqPlot/jqPlot#205, adding a new jqPlot global configuration option:

```diff
@@ -296,7 +296,7 @@
                 return window.G_vmlCanvasManager.initElement(canvas);
             }
 
-            var cctx = canvas.getContext('2d');
+            var cctx = canvas.getContext('2d', $.jqplot.config.canvasContextAttributes);
 
             var canvasBackingScale = 1;
             if (window.devicePixelRatio > 1 && (cctx.webkitBackingStorePixelRatio === undefined || 
@@ -376,7 +376,8 @@
         gapLength: 4,
         dotGapLength: 2.5,
         srcLocation: 'jqplot/src/',
-        pluginLocation: 'jqplot/src/plugins/'
+        pluginLocation: 'jqplot/src/plugins/',
+        canvasContextAttributes: { willReadFrequently: true }
     };
```

That also modifies line 299, but adds an extra option at line 379.

It looks like the line numbers are the more or less the same in the patch for `jqplot.core.js` and the concatenated `0-jqPlot.js` in PrimeFaces (quite possibly because `jqplot.core.js` would be the first concatenated file).

### Option 3

It might make sense to make that an option in `1-charts.js`, but I'm not sure. I'm afraid I don't know enough about the way PrimeFaces expect configuration options.
* Having a option that's scoped to particular plots would require more changes in jqPlot itself.
* The `$.jqplot.config.canvasContextAttributes` in my patch is global (in the sense it affects all jqPlot charts, since it's in the `$.jqplot.config` namespace). I don't know how PrimeFaces handles other options already in `$.jqplot.config`.

**Reply by melloware** [reply]
Thanks @harbulot i agree this should be easy but @imrankhurram-checkit seems to be struggling how to patch it but its just replacing the JS inside the JAR file.

**Reply by imrankhurram-checkit** [reply]
@melloware  I am afraid it is not that simple. I tried replacing chart.js with  1-charts.js and 0-jqPlot.js in my normal Primefaces jar file. Also added the fix on both lines about that canvasContextAttributes in 0-jqPlot.js but now the chart is not totally rendering and giving alot of errors in the console. e.g. 
Uncaught TypeError: Cannot read properties of undefined (reading 'eventListenerHooks')

Uncaught TypeError: Cannot set properties of undefined (setting 'CanvasOverlay')

Uncaught TypeError: Cannot read properties of undefined (reading 'postDrawHooks')

Uncaught TypeError: Cannot read properties of undefined (reading 'prototype')

and before all these errors I can also see 
Failed to load resource: the server responded with a status of 404 (Not Found)
Refused to execute script from '.../RES_NOT_FOUND' because its MIME type ('text/html') is not executable, and strict MIME type checking is enabled.

Do you guys have a fixed and working jar file for Primefaces 12 ?

**Reply by harbulot** [reply, marked as answer]
@imrankhurram-checkit Again, not a user of Primefaces, so can't comment much, but it looks like `0-jqPlot.js` and `1-charts.js` are in `primefaces-12.0.0-sources.jar` but not in `primefaces-12.0.0.jar`, where you only get a `charts.js` file instead.

To me, that suggests `0-jqPlot.js` and `1-charts.js` are processed into `charts.js` during the build.

Looking at `charts.js`, it looks indeed like a minified version of a concatenation of `0-jqPlot.js` and `1-charts.js`, which would be quite typical for bundling some JavaScript code.

I'm guessing that, if you've deleted `charts.js` and added `0-jqPlot.js` and `1-charts.js`, none of the pages that load `charts.js` will find it (hence the "not found" errors you see).

At first, you could just try to concatenate your modified `0-jqPlot.js` and `1-charts.js` together into a single `charts.js` file.
If you also want the file minified, there are a few JS obfuscation/minification tools out there, but bare in mind you probably don't want some function or option names mangled in the process (I'm not sure what PrimeFaces uses).

Failing all that, you could probably check out a copy of the 12.0.0 tag locally from the GitHub repository, changed those files, run the Maven build and either use that new jar file if you manage to get a successful build, or pick the generated `charts.js` to patch the official jar.

**Reply by melloware** [reply]
@harbulot has it 100% correct. They are named `0-` and `1-` because that is the order they are combined into `charts.js` by the build process.
